app/src/main/java/it/quip/android/model/Notification.java

```diff
@@ -7,79 +9,213 @@
+    }
+
+    public void setType(int type) {
+        this.mType = type;
+        this.safePutKey(PUSH_TYPE_KEY, type);
+    }
+
     public Notification() {
 
     }
 
-    public Notification(long uid, String text, String notificationImageUrl, int timestamp, boolean viewed) {
+    public static class with {
+        private final Context context;
+        private int notificationType;
+        private String notificationText;
+        private String notificationSenderId;
+        private String notificationReceiverId;
+        private String notificationImageUrl;
+
+        public with(Context c) {
+            this.context = c;
+            // defaults
+            this.notificationSenderId = null;
+            this.notificationReceiverId = null;
+            this.notificationImageUrl = null;
+            this.notificationType = Notification.STANDARD_NOTIFICATION;
+        }
+
+        public with type(int type) {
+            notificationType = type;
+            return this;
+        }
+
+        public with body(String bodyText) {
+            notificationText = bodyText;
+            return this;
+        }
+
+        public with sender(User user) {
+            notificationSenderId = user.getObjectId();
+            return this;
+        }
+
+        public with receiver(User user) {
+            notificationReceiverId = user.getObjectId();
+            return this;
+        }
+
+        public with imageUrl(String url) {
+            notificationImageUrl = url;
+            return this;
+        }
+
+        public Notification build() {
+            Notification n = new Notification(this);
+            n.mViewed = false;
+            return n;
+        }
+        public Notification deliver() {
+            Notification n = new Notification(this);
+            n.mViewed = false;
+            n.send();
+            return n;
+        }
+
+    }
+
+    private Notification(with builder) {
+        mType = builder.notificationType;
+        this.safePutKey(Notification.PUSH_TYPE_KEY, mType);
+        mSenderUid = builder.notificationSenderId;
+        this.safePutKey(Notification.PUSH_SENDER_ID, mSenderUid);
+        mReceiverUid = builder.notificationReceiverId;
+        this.safePutKey(Notification.PUSH_RECEIVER_ID, mReceiverUid);
+        mText = builder.notificationText;
+        this.safePutKey(Notification.PUSH_TEXT_BODY_KEY, mText);
+        mTimestamp = TimeUtils.currentTimestampInS();
+        this.safePutKey(Notification.PUSH_TIMESTAMP_KEY, mTimestamp);
+        mNotificationImageUrl = builder.notificationImageUrl;
+        this.safePutKey(Notification.PUSH_IMAGE_URL_KEY, mNotificationImageUrl);
+    }
 
-        this.uid = uid;
-        this.text = text;
-        this.notificationImageUrl = notificationImageUrl;
-        this.timestamp = timestamp;
-        this.viewed = viewed;
-        this.type = STANDARD_NOTIFICATION;
+    private void safePutKey(String key, Object value) {
```

Comment:
Let's move this into a shared BaseObject that all of our Parse model objects extend.